libtransmission/global-ip-cache.h

```diff
@@ -0,0 +1,123 @@
+    tr_global_ip_cache& operator=(tr_global_ip_cache const&) = delete;
+    tr_global_ip_cache& operator=(tr_global_ip_cache&&) = delete;
+
+    bool try_shutdown() noexcept;
+
+    [[nodiscard]] std::optional<tr_address> global_addr(tr_address_type type) const noexcept;
+    [[nodiscard]] std::optional<tr_address> global_source_addr(tr_address_type type) const noexcept;
+
+    [[nodiscard]] tr_address bind_addr(tr_address_type type) const noexcept;
+
+    void update_addr(tr_address_type type) noexcept;
+
+    // Whether this machine supports this IP protocol
+    array_ip_t<std::atomic_bool> has_ip_protocol_ = { true, true };
+
+private:
+    void set_global_addr(tr_address const& addr) noexcept;
+    void unset_global_addr(tr_address_type type) noexcept;
+    void set_source_addr(tr_address const& addr) noexcept;
+    void unset_addr(tr_address_type type) noexcept;
+
+    void start_timer(tr_address_type type, std::chrono::milliseconds msec) noexcept;
+    void stop_timer(tr_address_type type) noexcept;
+    [[nodiscard]] bool set_is_updating(tr_address_type type) noexcept;
+    void unset_is_updating(tr_address_type type) noexcept;
+
+    void update_global_addr(tr_address_type type) noexcept;
+    void update_source_addr(tr_address_type type) noexcept;
+
+    // Only use as a callback for web_->fetch()
+    void on_response_ip_query(tr_address_type type, tr_web::FetchResponse const& response) noexcept;
+
+    [[nodiscard]] static std::optional<tr_address> get_global_source_address(tr_address const& bind_addr, int& err_out);
+    [[nodiscard]] static std::optional<tr_address> get_source_address(
+        tr_address const& dst_addr,
+        tr_port dst_port,
+        tr_address const& bind_addr,
+        int& err_out);
+
+    tr_session const& session_;
+
+    enum class is_updating_t
+    {
+        NO,
+        YES,
+        ABORT
+    };
+    array_ip_t<is_updating_t> is_updating_ = {};
+    array_ip_t<std::mutex> is_updating_mutex_;
+    array_ip_t<std::condition_variable> is_updating_cv_;
+
+    // Never directly read/write IP addresses for the sake of being thread safe
+    // Use global_*_addr() for read, and set_*_addr()/unset_*_addr() for write instead
+    mutable array_ip_t<std::shared_mutex> global_addr_mutex_;
+    array_ip_t<std::optional<tr_address>> global_addr_;
+    mutable array_ip_t<std::shared_mutex> source_addr_mutex_;
+    array_ip_t<std::optional<tr_address>> source_addr_;
+
+    // Keep the timer at the bottom of the class definition so that it will be destructed first
+    // We don't want it to trigger after the IP addresses have been destroyed
+    // (The destructor will acquire the IP address locks before proceeding, but still)
+    array_ip_t<std::unique_ptr<libtransmission::Timer>> upkeep_timers_;
+
+    array_ip_t<std::atomic_size_t> ix_service_ = {};
+    static auto constexpr IPQueryServices = std::array{ "https://icanhazip.com"sv, "https://api64.ipify.org"sv };
```

Comment:
This could be moved to an anonymous namespace in `global-ip-cache.cc`, which would let us remove `using namespace std::literals` from the header

Reply:
adopted in dec109f22f7bfb3c4fd323136cf9d33100a0d424